airbyte-workers/src/main/java/io/airbyte/workers/process/ExitCodeWatcher.java

```diff
@@ -20,52 +20,101 @@
  * be able to retrieve the exit code).
  */
 @Slf4j
-public class ExitCodeWatcher implements Watcher<Pod> {
+public class ExitCodeWatcher implements ResourceEventHandler<Pod> {
 
+  private final String podName;
+  private final String podNamespace;
   private final Consumer<Integer> onExitCode;
-  private final Consumer<WatcherException> onWatchFailure;
-  private boolean exitCodeRetrieved = false;
+  private final Runnable onWatchFailure;
+  /**
+   * This flag is set to false when we either (a) find the pod's exit code, or (b) when the pod is
+   * deleted. This is so that we call exactly one of onExitCode and onWatchFailure, and we make that
+   * call exactly once.
+   */
+  private boolean active = true;
 
   /**
-   *
    * @param onExitCode callback used to store the exit code
    * @param onWatchFailure callback that's triggered when the watch fails. should be some failed exit
    *        code.
    */
-  public ExitCodeWatcher(final Consumer<Integer> onExitCode, final Consumer<WatcherException> onWatchFailure) {
+  public ExitCodeWatcher(final String podName,
+                         final String podNamespace,
+                         final Consumer<Integer> onExitCode,
```

Comment:
in practice aren't we running the following? that isn't storing the exit code (like the javadoc comment says), it is killing the pod, right?

```
       exitCode -> {
          fabricClient.pods().delete(pod);
        },
```

Reply:
the production case is to call `exitCodeFuture.complete(theExitCode)` - https://github.com/airbytehq/airbyte/pull/11861/files#diff-be3c5f5544310ca55f284848493a7afe72533bd82a737a044d3f58e9c82d71abR539

This delete thing is happening in the integration test, where there are actually 2 watchers: One from the KubePodProcess, populating the normal exit code future; and one from the test itself, which just deletes the pod as soon as it exits (to test that we still capture the exit code even if it's deleted very quickly).